src/main/java/net/dv8tion/jda/api/utils/AttachedFile.java

```diff
@@ -0,0 +1,183 @@
+     * <br>This is primarily used for message edit requests, to specify which attachments to retain in the message after the update.
+     *
+     * @param  id
+     *         The id of the attachment to retain
+     *
+     * @throws IllegalArgumentException
+     *         If the id is not a valid snowflake
+     *
+     * @return {@link AttachmentUpdate}
+     */
+    @Nonnull
+    static AttachmentUpdate fromAttachment(@Nonnull String id)
+    {
+        return AttachmentUpdate.fromAttachment(id);
+    }
+
+
+    /**
+     * Creates an {@link AttachmentUpdate} with the given attachment.
+     * <br>This is primarily used for message edit requests, to specify which attachments to retain in the message after the update.
+     *
+     * @param  attachment
+     *         The attachment to retain
+     *
+     * @return {@link AttachmentUpdate}
+     */
+    @Nonnull
+    static AttachmentUpdate fromAttachment(@Nonnull Message.Attachment attachment)
+    {
+        return AttachmentUpdate.fromAttachment(attachment);
+    }
+
+    /**
+     * Used internally to build the multipart request.
+     *
+     * @param builder
+     *        The {@link MultipartBody.Builder} used for the request body
+     * @param index
+     *        The index of the attachment, ignored for {@link AttachmentUpdate}
+     */
+    default void addPart(MultipartBody.Builder builder, int index) {}
+
+    /**
+     * Used internally to build attachment descriptions for requests.
+     * <br>This contains the id/index of the attachment, and the name of the file.
+     *
+     * @param  index
+     *         The reference index (should be same as {@link #addPart(MultipartBody.Builder, int)})
+     *
+     * @return {@link DataObject} for the attachment
+     */
+    @Nonnull
+    DataObject toAttachmentData(int index);
+
+    /**
+     * Build a complete request using the provided files and payload data.
+     * <br>If the provided {@code payloadJson} is null, the multipart request will not set {@code attachments}.
+     *
+     * @param  files
+     *         The files to upload/edit
+     * @param  payloadJson
+     *         The payload data to send, excluding {@code attachments} field
+     *
+     * @throws IllegalArgumentException
+     *         If the file list is null
+     *
+     * @return {@link MultipartBody.Builder}
+     */
+    @Nonnull
+    static MultipartBody.Builder createMultipartBody(@Nonnull List<? extends AttachedFile> files, @Nullable DataObject payloadJson)
+    {
+        MultipartBody.Builder builder = new MultipartBody.Builder().setType(MultipartBody.FORM);
+        DataArray descriptors = DataArray.empty();
+        for (int i = 0; i < files.size(); i++)
+        {
+            AttachedFile file = files.get(i);
+            file.addPart(builder, i);
+            descriptors.add(file.toAttachmentData(i));
+        }
+
+        if (payloadJson == null)
+            return builder;
+        payloadJson.put("attachments", descriptors);
```

Comment:
Because setting them in the json, removes all unset attachments from the message.

Reply:
This is because the files provided could be an empty list due to none being added, so edits would send `payload_json.attachments: []` which would remove them all.

Gotcha.